core/services/gateway/handlers/functions/subscriptions.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error during functions_router.NewFunctionsRouter: %s", err)
+	}
+	return &onchainSubscriptions{
+		config:             config,
+		subscriptions:      make(map[common.Address]*functions_router.IFunctionsSubscriptionsSubscription),
+		client:             client,
+		router:             router,
+		blockConfirmations: big.NewInt(int64(config.BlockConfirmations)),
+		lggr:               lggr.Named("OnchainSubscriptions"),
+		stopCh:             make(utils.StopChan),
+	}, nil
+}
+
+func (s *onchainSubscriptions) Start(ctx context.Context) error {
+	return s.StartOnce("OnchainSubscriptions", func() error {
+		s.lggr.Info("starting onchain subscriptions")
+		if s.config.QueryFrequencySec == 0 {
+			return errors.New("OnchainSubscriptionsConfig.UpdateFrequencySec must be greater than 0")
+		}
+		if s.config.QueryTimeoutSec == 0 {
+			return errors.New("OnchainSubscriptionsConfig.UpdateTimeoutSec must be greater than 0")
+		}
+		if s.config.QueryRangeSize == 0 {
+			return errors.New("OnchainSubscriptionsConfig.QueryRangeSize must be greater than 0")
+		}
+
+		s.closeWait.Add(1)
+		go s.queryLoop()
+
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) Close() error {
+	return s.StopOnce("OnchainSubscriptions", func() (err error) {
+		s.lggr.Info("closing onchain subscriptions")
+		close(s.stopCh)
+		s.closeWait.Wait()
+		return nil
+	})
+}
+
+func (s *onchainSubscriptions) GetSubscription(address common.Address) *functions_router.IFunctionsSubscriptionsSubscription {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	subscription, ok := s.subscriptions[address]
+	if !ok {
+		return nil
+	}
+	return subscription
+}
+
+func (s *onchainSubscriptions) queryLoop() {
+	defer s.closeWait.Done()
+
+	ticker := time.NewTicker(time.Duration(s.config.QueryFrequencySec) * time.Second)
+	defer ticker.Stop()
+
+	var start uint64 = 1
+
+	queryFunc := func() {
+		ctx, cancel := utils.ContextFromChanWithTimeout(s.stopCh, time.Duration(s.config.QueryTimeoutSec)*time.Second)
+		defer cancel()
+
+		latestBlockHeight, err := s.client.LatestBlockHeight(ctx)
+		if err != nil || latestBlockHeight == nil {
+			s.lggr.Errorw("Error calling LatestBlockHeight", "err", err, "latestBlockHeight", latestBlockHeight.Int64())
+			return
+		}
+
+		blockNumber := big.NewInt(0).Sub(latestBlockHeight, s.blockConfirmations)
+
+		count, err := s.getSubscriptionsCount(ctx, blockNumber)
+		if err != nil {
+			s.lggr.Errorw("Error getting subscriptions count", "err", err)
+			return
+		}
+		if count == 0 {
+			s.lggr.Info("Router has no subscriptions yet")
+			return
+		}
+
+		if start > count {
+			start = 1
+		}
+
+		end := start + uint64(s.config.QueryRangeSize)
+		if end > count {
+			end = count
+		}
+		if err := s.querySubscriptionsRange(ctx, blockNumber, start, end); err != nil {
+			s.lggr.Errorw("Error querying subscriptions", "err", err, "start", start, "end", end)
+			return
+		}
+
+		start = end + 1
+	}
+
+	for {
```

Comment:
Call it once before the for loop so node doesn't need to always wait for QueryFrequencySec on startup.